lib/flask_tasks.py

```diff
@@ -0,0 +1,194 @@
+#
+# mostly copied from https://github.com/miguelgrinberg/flack/commit/0c372464b341a2df60ef8d93bdca2001009a42b5?diff=unified
+# video https://www.youtube.com/watch?v=tdIIJuPh3SI&feature=youtu.be
+#
+from functools import wraps
+import threading
+import time
+import uuid
+import logging
+import sys
+import traceback
+import json
+
+from flask import Blueprint, abort, current_app, g, request, jsonify
+from werkzeug.exceptions import HTTPException, InternalServerError
+from flask import url_for as _url_for, current_app, _request_ctx_stack
+
+logging.basicConfig(format='%(asctime)s | %(name)s | %(message)s',
+                    level=logging.DEBUG)
+
+
+tasks_bp = Blueprint('tasks', __name__)
+tasks = {}
+testing = False
+
+
+def timestamp():
+    """Return the current timestamp as an integer."""
+    return int(time.time())
+
+
+def url_for(*args, **kwargs):
+    """
+    url_for replacement that works even when there is no request context.
+    """
+    if '_external' not in kwargs:
+        kwargs['_external'] = False
+    reqctx = _request_ctx_stack.top
+    if reqctx is None:
+        if kwargs['_external']:
+            raise RuntimeError('Cannot generate external URLs without a '
+                               'request context.')
+        with current_app.test_request_context():
+            return _url_for(*args, **kwargs)
+    return _url_for(*args, **kwargs)
+
+
+@tasks_bp.before_app_first_request
+def before_first_request():
+    """Start a background thread that cleans up old tasks."""
+    def clean_old_tasks():
+        logging.debug("Cleanup old tasks")
+        """
+        This function cleans up old tasks from our in-memory data structure.
+        """
+        global tasks
+        while True:
+            # Only keep tasks that are running or that finished less than 5
+            # minutes ago.
+            five_min_ago = timestamp() - 5 * 60
+            tasks = {id: task for id, task in tasks.items()
+                     if 't' not in task or task['t'] > five_min_ago}
+            time.sleep(60)
+
+    if not testing:
+        logging.debug("Production mode, run cleanups")
+        thread = threading.Thread(target=clean_old_tasks)
+        #thread.start()
+    else:
+        logging.debug("Testing mode, no cleanup!")
+
+
+def async_task(f):
+    """
+    This decorator transforms a sync route to asynchronous by running it
+    in a background thread.
+    """
+    @wraps(f)
+    def wrapped(*args, **kwargs):
+        def task(app, environ):
+            # Create a request context similar to that of the original request
+            # so that the task can have access to flask.g, flask.request, etc.
+            with app.request_context(environ):
+                try:
+                    # Run the route function and record the response
+                    logging.debug("____________________________")
```

Comment:
Maybe you better create a dedicated logging method for drawing those lines automatically.

Reply:
Just deleted it